src/storage.rs

```diff
@@ -1,296 +1,494 @@
+            self.uninitialized -= 1;
+        } else {
+            // We can safely drop this, it's initialized.
+            *self.data.get_unchecked_mut(cursor) = elem;
+        }
+    }
+
+    /// `cursor` is the first position that gets moved to the back,
+    /// free memory will be created between `cursor - 1` and `cursor`.
+    unsafe fn grow(&mut self, cursor: usize, by: usize) {
+        assert!(by >= self.data.len());
+
+        // Calculate how many elements we need to move
+        let to_move = self.data.len() - cursor;
+
+        // Reserve space and set the new length
+        self.data.reserve_exact(by);
+        let new = self.data.len() + by;
+        self.data.set_len(new);
+
+        // Move the elements after the cursor to the end of the buffer.
+        // Since we grew the buffer at least by the old length,
+        // the elements are non-overlapping.
+        let src = self.data.as_ptr().offset(cursor as isize);
+        let dst = self.data.as_mut_ptr().offset((cursor + by) as isize);
+        ptr::copy_nonoverlapping(src, dst, to_move);
+
+        self.uninitialized += by;
+    }
+
+    /// Called when dropping the ring buffer.
+    unsafe fn clean(&mut self, cursor: usize) {
+        let mut cursor = CircularIndex::new(cursor, self.data.len());
+        let end = cursor - 1;
+
+        while let Some(i) = cursor.step(end) {
+            if self.uninitialized > 0 {
+                self.uninitialized -= 1;
+            } else {
+                ptr::drop_in_place(self.data.get_unchecked_mut(i) as *mut T);
+            }
+        }
+
+        self.data.set_len(0);
+    }
+
+    #[cfg(test)]
+    fn num_initialized(&self) -> usize {
+        self.data.len() - self.uninitialized
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+struct Reader {
+    generation: usize,
+    last_index: usize,
+}
+
+impl Reader {
+    fn set_inactive(&mut self) {
+        self.last_index = !0;
+    }
+
+    fn active(&self) -> bool {
+        self.last_index != !0
+    }
+
+    fn distance_from(&self, last: CircularIndex, current_gen: usize) -> usize {
+        let this = CircularIndex {
+            index: self.last_index,
+            size: last.size,
+        };
+
+        match this - last.index {
+            0 if self.generation == current_gen => last.size,
+            x => x,
+        }
+    }
+
+    fn needs_shift(&self, last_index: usize, current_gen: usize) -> bool {
+        self.last_index > last_index
+            || (self.last_index == last_index && self.generation != current_gen)
+    }
+}
+
+/// The reader id is used by readers to tell the storage where the last read ended.
 #[derive(Debug)]
-struct InternalReaderId {
-    written: usize,
-    index: usize,
-    alive: Arc<AtomicBool>,
+pub struct ReaderId<T: 'static> {
+    id: usize,
+    marker: PhantomData<&'static [T]>,
+    // stupid way to make this `Sync`,
+    // never actually locked
+    drop_notifier: Mutex<Sender<usize>>,
```

Comment:
Comment ^

Reply:
Alternatively you could just `unsafe impl Sync` for this

Reply:
That's what there was before, I don't know why it was changed.

Reply:
same as below